fax/implicit/twophase.py

```diff
@@ -32,98 +29,44 @@ def convergence_test(x_new, x_old):
         )
 
         return sol
-    return _default_solver
+    return _default_solve
 
 
-def make_adjoint_fixed_point_iteration(
-        param_func, default_rtol=1e-4, default_atol=1e-4, default_max_iter=5000,
-        default_batched_iter_size=1):
+@functools.partial(jax.custom_vjp, nondiff_argnums=(0, 1, 3))
+def two_phase_solve(param_func, init_xs, params, solvers=()):
+    if solvers:
+        fwd_solver = solvers[0]
+    else:
+        fwd_solver = default_solver(param_func)
 
-    # define a flat function to fit the vjp API
-    def flat_func(i, x, params):
-        return param_func(params)(i, x)
+    return fwd_solver(init_xs, params)
 
-    def _adjoint_iteration_vjp(g, ans, init_xs, params):
-        dvalue = g
-        del init_xs
-        init_dxs = dvalue
 
-        fp_vjp_fn = jax.vjp(partial(flat_func, ans.iterations),
-                            ans.value, params)[1]
+def two_phase_fwd(param_func, init_xs, params, solvers):
+    sol = two_phase_solve(param_func, init_xs, params, solvers)
+    return sol, (sol.value, params)
 
-        def dfp_fn(i, dout):
-            del i
+
+def two_phase_rev(param_func, init_xs, solvers, res, sol_bar):
+    del init_xs
+
+    def param_dfp_fn(packed_params):
+        v, p, dvalue = packed_params
+        _, fp_vjp_fn = jax.vjp(lambda x: param_func(p)(x), v)
+
+        def dfp_fn(dout):
             dout = fp_vjp_fn(dout)[0] + dvalue
             return dout
 
-        rtol, atol = converge.adjust_tol_for_dtype(
-            default_rtol, default_atol, init_dxs.dtype)
+        return dfp_fn
 
-        def convergence_test(x_new, x_old):
-            return converge.max_diff_test(x_new, x_old, rtol, atol)
+    value, params = res
+    dsol = two_phase_solve(param_dfp_fn,
+                           sol_bar.value,
+                           (value, params, sol_bar.value),
+                           solvers[1:])
+    _, dparam_vjp = jax.vjp(lambda p: param_func(p)(value), params)
+    return dparam_vjp(dsol.value)
 
-        dsol = loop.fixed_point_iteration(
-            init_x=init_dxs,
-            func=dfp_fn,
-            convergence_test=convergence_test,
-            max_iter=default_max_iter,
-            batched_iter_size=default_batched_iter_size,
-        )
 
-        return fp_vjp_fn(dsol.value)[1], dsol
-
-    return _adjoint_iteration_vjp
-
-
-def two_phase_solver(param_func, forward_solver=None, default_rtol=1e-4,
-                     default_atol=1e-4, default_max_iter=5000,
-                     default_batched_iter_size=1):
-    """ Create an implicit function of the parameters and define its VJP rule.
```

Comment:
This was a great docstring.  You didn't want to keep most of it?

Reply:
Yes! I plan on bringing that back but I wanted to get some comments on this first since the docs will have to be changed a bit depending on what the final api ends up being.